### DISCUSSION TITLE
How to specify `ssl_mode` in DATABASE_URL for MySQL

### DISCUSSION THREAD

**Discussion by amsharma9** [diesel-rs/diesel#3633] [Q&A] [answered]
I am new to Rust. I have specified the following in my `.env` file
DATABASE_URL=mysql://myusr:mypass@localhost/pubdb?ssl_mode=DISABLED

but I get an error written below:
D:\Rust\diesel_demo>diesel setup
The --database-url argument must be passed, or the DATABASE_URL environment variable must be set.

The link: https://docs.diesel.rs/2.0.x/diesel/mysql/struct.MysqlConnection.html#method.establish shows a "*" at the end of `ssl_mode` and before the next `&`. What is that? Am I specifying it the right way?

I have tried all of the below, all of them don't work. The the first one gives `SSL error` and the other two give same error as above.
DATABASE_URL=mysql://myusr:mypass@localhost/pubdb

DATABASE_URL=mysql://myusr:mypass@localhost/pubdb?ssl_mode=DISABLED*
DATABASE_URL="mysql://myusr:mypass@localhost/pubdb?ssl_mode=DISABLED"

**Comment by Mingun** [comment]
Probably the star means `SSL_MODEDISABLED` for your case. Or `SSL_MODE_DISABLED`

**Reply by amsharma9** [reply]
Are these valid values for `ssl_mode` @Mingun? The documentation doesn't mention them. I see only `DISABLED`.

**Reply by weiznich** [reply]
Again: Its not related. The database url that works from the command line is correct. There is **no** additional processing done for database URL's from `.env` files. Diesel does not even see a difference between those coming from `.env` and normal environment variables.

**Reply by Mingun** [reply]
That was just a blindly guess of what the star could mean. Looking at the documentation I think now that this is just a excess character, because `&` and `*` symbols on the keyboard are pretty close to each other (numbers 7 and 8 specifically)

**Comment by weiznich** [comment]
The error message indicates that this seem not to be related to `ssl_mode` at all. For some reason diesel cli fails to pickup the `.env` file in your environment. That's likely highly dependent on your environment so it's hard to really tell what's wrong there. 
Can you make sure that you saved the `.env` file using an UTF-8 encoding? Also can you make sure that the file is really in the current directory where you are trying to execute `diesel setup`?

**Reply by amsharma9** [reply]
The `.env` file is in the project's home dir, i.e. D:\Rust\diesel_demo. All other files like diesel.toml, cargo.toml are in this directory itself. If diesel was not reading the file then how come the error changes when `?ssl_mode=DISABLED` part is removed from this file.
Secondly, the below runs successfully from command line but not when we use `.env` file.
```
D:\Rust\diesel_demo>diesel setup --database-url=mysql://myusr:mypass@localhost/pubdb?ssl_mode=DISABLED
Creating database: pubdb
```

**Reply by weiznich** [reply]
To write that again: If the database url works when you pass it directly it's not an issue with your database connection but with your `.env` file. That means the file is not read for whatever reason. I do not know how you created that file or how you edited the file but one possibility would be that you created the file via a shell command `echo DATABASE_URL=…>.env` and edited it with some real editor. That alone might change the file encoding from something else to UTF-8, which is understood by the crate responsible for resolving `.env` files.

**Reply by amsharma9** [reply]
I used VSCode to create the file. Not sure if it's in UTF-8 encoding. Any suggestions on how I can confirm that. Another info I wanted to add, the following command doesn't work from command line. See the error reported back:
```
D:\Rust\diesel_demo>diesel setup --database-url=mysql://myusr:mypass@localhost/pubdb
Creating database: pubdb
SSL connection error: unknown error number
```
My guess is that Diesel is probably not able to interpret parameters after `?` when using `.env` file. I have confirmed that `.env` file is being read. If I specify below in the file I get a panic error:
```
DATABASE_URL=
```
```
D:\Rust\diesel_demo>diesel setup
thread 'main' panicked at '`` is not a valid database URL. It should start with `mysql://`, or maybe you meant to use the `sqlite` feature which is not enabled.', C:\Users\MESPL\.cargo\registry\src\github.com-1ecc6299db9ec823\diesel_cli-2.0.1\src\database.rs:83:17
stack backtrace:
   0:     0x7ff6073023d2 - <unknown>
   1:     0x7ff60731bd0b - <unknown>
....
....
```

**Reply by weiznich** [reply]
No, the content of that file does not change how it is read. I would guess its either a encoding issue or your .env file is otherwise incompatible with the dotenvy crate (which is internally used.). If you suspect that this is a bug you really need to provide an reproducible example. That includes the `.env`  file thats causing your issue

> Not sure if it's in UTF-8 encoding

VsCode should show that somewhere in the bottom bar.

> Another info I wanted to add, the following command doesn't work from command line. See the error reported back

Thats not surprising if you need to manually set the ssl mode. Again thats repeated to your local setup and not to diesel

> My guess is that Diesel is probably not able to interpret parameters after ? when using .env file.

Again thats not the case. Diesel does not do any processing there.

Thats likely my last answer here as I cannot contribute anything meaningful to this discussion based on the shared information.

**Reply by amsharma9** [reply]
Thanks for your responses @weiznich. I hope we can identify and resolve the issue. Vscode and Notepad++ show UTF-8 as the type for `.env`. I have attached the current file with all the options that I have tried. Had to rename it to `env.txt` to upload. If you need anything else to recreate the issue, kindly LMK, I will share.

Pardon my lack of knowledge. Could the MySQL library be an issue. I had to struggle to install it. Diesel documentation says it wants - `libmysqlclient` for the Mysql backend. I have `mysqlclient-sys v0.2.5` showing up during compilation. Do they serve the same purpose?

[env.txt](https://github.com/diesel-rs/diesel/files/11523744/env.txt)

**Reply by weiznich** [reply, marked as answer]
This `.env` file is invalid, as it contains lines that are not accepted by the crate we use to instance such files (`dotenvy`).  This causes the complete file to be ignored. Especially lines starting with `;` cause problems. I've opened #3636 to emit a error in such cases.

A working `.env` file looks like:
```
DATABASE_URL=mysql://myusr:system@localhost/pubdb?ssl_mode=DISABLED
```
(At least that database url is then used by my `diesel-cli` version)

**Reply by amsharma9** [reply]
Sorry for the late response. It would be good for developers if you can emit an error when you face invalid `.env`.
Just wanted to confirm if unix style `#` works to comment lines in `.env`. Thanks.

**Reply by weiznich** [reply]
> It would be good for developers if you can emit an error when you face invalid .env.

That's already implemented with #3636 and part of diesel-cli 2.1
